transaction/std/src/input_materials.rs

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) 2022 The MobileCoin Foundation
+
+//! Input Materials is a helper struct for the transaction builder.
+//! The transaction builder can get inputs either from input credentials,
+//! or signed contingent inputs. In one case the input is being signed by us,
+//! and in the other it is already signed.
+//!
+//! The transaction builder is required to sort these in order of tx public key
+//! of the first element of the ring being signed, then hand them off to build
+//! the actual signatures. This enum makes it convenient to do this.
+
+use crate::InputCredentials;
+use mc_crypto_keys::CompressedRistrettoPublic;
+use mc_transaction_core::{ring_signature::InputRing, tx::TxIn, Amount, SignedContingentInput};
+
+/// Material that can be used by the transaction builder to create an input to
+/// a transaction.
+#[derive(Debug, Clone)]
+pub enum InputMaterials {
+    /// Signable input materials
+    Signable(InputCredentials),
+    /// Presigned input materials
+    Presigned(SignedContingentInput),
+}
+
+impl From<InputCredentials> for InputMaterials {
+    fn from(src: InputCredentials) -> Self {
+        Self::Signable(src)
+    }
+}
+
+impl From<SignedContingentInput> for InputMaterials {
+    fn from(src: SignedContingentInput) -> Self {
+        Self::Presigned(src)
+    }
+}
+
+impl InputMaterials {
+    /// Get the sort key for whichever type of input this is
+    pub fn sort_key(&self) -> &CompressedRistrettoPublic {
+        match self {
+            InputMaterials::Signable(cred) => &cred.ring[0].public_key,
+            InputMaterials::Presigned(input) => &input.tx_in.ring[0].public_key,
+        }
+    }
+
+    /// Get the amount for whichever type of input this is
+    pub fn amount(&self) -> Amount {
+        match self {
+            InputMaterials::Signable(cred) => cred.input_secret.amount,
+            InputMaterials::Presigned(input) => Amount {
+                value: input.pseudo_output_amount.value,
+                token_id: input.pseudo_output_amount.token_id.into(),
+            },
+        }
+    }
+
+    /// Get the ring size for whichever type of input this is
+    pub fn ring_size(&self) -> usize {
+        match self {
+            InputMaterials::Signable(cred) => cred.ring.len(),
+            InputMaterials::Presigned(input) => input.tx_in.ring.len(),
+        }
+    }
+}
+
+// Helper which converts from InputMaterials (TransactionBuilder type) to
+// InputRing (rct_bulletproofs type)
+impl From<InputMaterials> for InputRing {
```

Comment:
Would it make sense to make this `From<&InputMaterials>` (by ref) to match the impl below?

Reply:
So like, I am somewhat torn about this. We only need the `From<InputMaterials>` in real life, and coding it that way avoids some unnecessary copies.

There are other places in our code where we choose to implement `From<&_>` instead. But I wonder if it's not more idiomatic to implement `From<_>` and then clone before converting only if necessary. We consistently do `From<_>` conversions whenever error objects are being converted, because otherwise the `?` operator doesn't work.

I'm inclined to only implement the version I actually need, and also, to make it possible for developers to avoid unnecessary copies / dynamic memory allocations, as a rule of thumb.

Reply:
I think that `TryFrom<&>` makes total sense though, because usually if the conversion is invalid you don't want to consume the argument.